### PR TITLE
Adding RAABBVI implementation

### PR DESCRIPTION
**Pull request by Manushi22**

### CODE REVIEW

File: viabel/objectives.py [hunk 1]
```diff
@@ -133,6 +141,10 @@ def variational_objective(var_param):
             return -lower_bound
         self._objective_and_grad = value_and_grad(variational_objective)
 
+    def _hessian_vector_product(self, var_param, x):
+        hvp = make_hvp(self.variational_objective)(var_param)[0]
```
Comment (jhuggins) [inline viabel/objectives.py:145]: You shouldn't need to call `make_hvp` every time. Just do it once in `__init__` and then reuse

File: docs/source/api/api.rst [hunk 2]
```diff
@@ -70,3 +70,4 @@ Optimization
    RMSProp
    AdaGrad
    FASO
+   RAABBVI
```
Comment (jhuggins) [inline docs/source/api/api.rst:73]: You should also add the other base optimizers like AveragedRMSProp

**Review by jhuggins — CHANGES_REQUESTED** [review]
A couple minor comments. Please make sure you are able to build the documentation too

File: viabel/optimization.py [hunk 3]
```diff
@@ -142,6 +144,19 @@ def descent_direction(self, grad):
 
 class RMSProp(StochasticGradientOptimizer):
     """RMSProp optimization method
+    
+    Parameters
+    -----------
+    beta : `float` optional
+        hyper-parameter. The default is 0.9
```
Comment (jhuggins) [inline viabel/optimization.py:151]: It would be good to describe the algorithm in the comment so the beta hyperparameter makes sense. Same with the other optimization classes

File: viabel/optimization.py [hunk 4]
```diff
@@ -151,6 +166,11 @@ def __init__(self, learning_rate, *, weight_decay=0, beta=0.9, jitter=1e-8,
         super().__init__(learning_rate, weight_decay=weight_decay, diagnostics=diagnostics)
 
     def reset_state(self):
+        """
+        resetting the state
```
Comment (jhuggins) [inline viabel/optimization.py:170]: Please describe in more detail what is actually done – that is, what does "state" actually mean here?

File: viabel/optimization.py [hunk 5]
```diff
@@ -165,8 +185,20 @@ def descent_direction(self, grad):
         return descent_dir
 
 
-class WindowedAdaGrad(StochasticGradientOptimizer):
+class WindowedAdagrad(StochasticGradientOptimizer):
     """Adam optimization method
```
Comment (jhuggins) [inline viabel/optimization.py:189]: Description needs to be updated. Definitely describe how the descent direction is calculated since it is non-standard

File: viabel/optimization.py [hunk 6]
```diff
@@ -176,6 +208,9 @@ def __init__(self, learning_rate, *, weight_decay=0, window_size=10, jitter=1e-8
         super().__init__(learning_rate, weight_decay=weight_decay, diagnostics=diagnostics)
 
     def reset_state(self):
+        """
+        resetting the state
```
Comment (jhuggins) [inline viabel/optimization.py:212]: again, state specifically what is getting reset

File: viabel/optimization.py [hunk 7]
```diff
@@ -187,15 +222,28 @@ def descent_direction(self, grad):
...
+        Descent direction of the optimization algorithm
     """
 
     def __init__(self, learning_rate, *, weight_decay=0, jitter=1e-8, diagnostics=False):
         self._jitter = jitter
         super().__init__(learning_rate, weight_decay=weight_decay, diagnostics=diagnostics)
 
     def reset_state(self):
+        """
+        resetting the state
```
Comment (jhuggins) [inline viabel/optimization.py:245]: same here

File: viabel/optimization.py [hunk 8]
```diff
@@ -213,6 +273,10 @@ def __init__(self, learning_rate, *, jitter=1e-8,
         super().__init__(learning_rate, diagnostics=diagnostics)
 
     def reset_state(self):
+        """
+        resetting the state
```
Comment (jhuggins) [inline viabel/optimization.py:277]: same here

File: viabel/optimization.py [hunk 9]
```diff
@@ -245,6 +323,9 @@ def __init__(self, learning_rate, *, beta_1=0.9, jitter=1e-8,
...
```
Comment (jhuggins) [inline viabel/optimization.py:327]: same here

File: viabel/stan_models/weighted_lin_regression.stan [hunk 10]
```diff
@@ -0,0 +1,26 @@
+data {
```
Comment (jhuggins) [inline viabel/stan_models/weighted_lin_regression.stan:1]: add a comment at the start summarizing what this model is used for

File: viabel/stan_models/weighted_lin_regression_sgd.stan [hunk 11]
```diff
@@ -0,0 +1,28 @@
+data {
```
Comment (jhuggins) [inline viabel/stan_models/weighted_lin_regression_sgd.stan:1]: add a comment at the start summarizing what this model is used for

File: viabel/optimization.py [hunk 12]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
                     iterate_average_history=np.array(iterate_average_history),
                     descent_dir_history=np.array(descent_dir_history),
                     ess_and_mcse_k_history=np.array(ess_and_mcse_k_history),
                     ess_history=np.array(ess_history),
                     mcse_history=np.array(mcse_history)
                     )
+                    
+
+class RAABBVI(FASO):
+    """A robust, automated, and accurate BBVI optimizer
```
Comment (jhuggins) [inline viabel/optimization.py:527]: Add a short description of how it works, plus a point to the preprint

File: viabel/optimization.py [hunk 13]
```diff
@@ -185,23 +222,141 @@ def descent_direction(self, grad):
...
+        else:
+            descent_dir = momentum / np.sqrt(self._jitter+np.sum(avg_grad_sq))
+        self._momentum = momentum
+        self._avg_grad_sq = avg_grad_sq
+        self._t = t
+        return descent_dir
+        
+        
 class FASO(Optimizer):
     """Fixed-learning rate stochastic optimization meta-algorithm
```
Comment (jhuggins) [inline viabel/optimization.py:361]: Add a short description of how it works, plus a pointer to the preprint

File: viabel/optimization.py [hunk 14]
```diff
@@ -238,7 +393,7 @@ def __init__(self, sgo, *, mcse_threshold=0.1, W_min=200, ESS_min=None,
             raise ValueError('"ESS_min" must be greater than zero')
 
     def optimize(self, n_iters, objective, init_param):
-        dim = init_param.size
+        dim = int(init_param.size/2)
```
Comment (jhuggins) [inline viabel/optimization.py:396]: this doesn't seem like a good way to determine the dimension as it only works for MF Gaussian family

File: viabel/optimization.py [hunk 15]
```diff
@@ -185,23 +222,141 @@ def descent_direction(self, grad):
         return descent_dir
 
 
-class AdaGrad(StochasticGradientOptimizer):
+class Adagrad(StochasticGradientOptimizer):
     """Adagrad optimization method
```
Comment (jhuggins) [inline viabel/optimization.py:226]: describe the algorithm

File: viabel/optimization.py [hunk 16]
```diff
@@ -185,23 +222,141 @@ def descent_direction(self, grad):
...
+        """
         self._sum_grad_sq = 0
 
     def descent_direction(self, grad):
         self._sum_grad_sq += grad**2
         descent_dir = grad / np.sqrt(self._jitter + self._sum_grad_sq)
         return descent_dir
 
+class AveragedRMSProp(StochasticGradientOptimizer):
+    """Averaged RMSProp optimization method
```
Comment (jhuggins) [inline viabel/optimization.py:255]: describe the algorithm

File: viabel/optimization.py [hunk 17]
```diff
@@ -185,23 +222,141 @@ def descent_direction(self, grad):
...
+        if self._component_wise:
+            descent_dir = grad / np.sqrt(self._jitter+avg_grad_sq)
+        else:
+            descent_dir = grad / np.sqrt(self._jitter+np.sum(avg_grad_sq))
+        self._avg_grad_sq = avg_grad_sq
+        self._t = t
+        return descent_dir
+        
+class AveragedAdam(StochasticGradientOptimizer):
+    """Averaged Adam optimization method
```
Comment (jhuggins) [inline viabel/optimization.py:302]: describe the algorithm

File: viabel/optimization.py [hunk 18]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+        kappa_sample_hist = []
+        k_new_hist = [0]
+        k_conv_hist = []
+        k_Rhat_hist = []
+        stopped = False
+        convg_iters_hist = []
+        predicted_iters_hist = []
+        stopping_crt = []
+        k_stopped_final_hist = []
+        LR = [] #learning rate for WLS regression
```
Comment (jhuggins) [inline viabel/optimization.py:720]: use a more descriptive variable name

File: viabel/optimization.py [hunk 19]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+                    opt = super().optimize(K_max, objective, iterate_average_curr)
+                if opt['k_stopped'] is not None and k!=0: #removing the number of convergence iterations of initial learning rate
+                    convg_iters = opt['k_stopped']
+                    convg_iters_hist.append(convg_iters)
+                    CI.append(convg_iters)
+                iterate_average_curr = opt['opt_param']
+                iterate_average_curr_hist.append(iterate_average_curr)
+                k_new = opt['k_stopped']
+                
+                if opt['k_Rhat'] is not None and k_new is not None:
```
Comment (jhuggins) [inline viabel/optimization.py:740]: comments explaining each of these if/else statements would be helpful

File: viabel/optimization.py [hunk 20]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+        k_add = 0 #Iteration number when convergence reached for a fixed step size
+        k_stopped_final = None #Iteration number when stopping rule criteria met
+        sgo = self._sgo
+        diagnostics = self._sgo._diagnostics
+        if isinstance(self._sgo, AveragedRMSProp) or isinstance(self._sgo, AveragedAdam):
+            reg_model = StanModel_cache(model_name='weighted_lin_regression_sgd')
+        else:
+            reg_model = StanModel_cache(model_name='weighted_lin_regression')
+        iterate_average_curr = init_param.copy()
+        iterate_average_curr_hist = [iterate_average_curr]
```
Comment (jhuggins) [inline viabel/optimization.py:695]: a comment or summarizing when / how all of these "history" lists are used would be helpful, especially if they are not used uniformly. Why are some `None` and some `[]` to start?

In fact, would it be possible to use a dictionary from the start to hold all of these histories, since that's how they get returned? You could use a `defaultdict` with default type `list`

File: viabel/optimization.py [hunk 21]
```diff
@@ -247,6 +402,7 @@ def optimize(self, n_iters, objective, init_param):
         variational_param = init_param.copy()
```
Comment (jhuggins) [inline viabel/optimization.py:402]: same comment about this history variables applies to FASO

File: viabel/optimization.py [hunk 22]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+                
+                if opt['k_Rhat'] is not None and k_new is not None:
+                    k_Rhat_hist.append(opt['k_Rhat']+k_add)
+                else:
+                    k_Rhat_hist.append(opt['k_Rhat'])
+                    
+                if opt['k_conv'] is not None and k_new is not None:
+                    k_conv_hist.append(opt['k_conv']+k_add)
+                else:
+                    k_conv_hist.append(opt['k_conv'])
```
Comment (jhuggins) [inline viabel/optimization.py:748]: so are you appending `None` here?

File: viabel/optimization.py [hunk 23]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+                
+                if k_new is not None:
+                    k_new_hist.append(k_new+k_add)   
+                else:
+                    k_new_hist.append(k_new)
```
Comment (jhuggins) [inline viabel/optimization.py:753]: also appending `None` here?

File: viabel/optimization.py [hunk 24]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+                    #k_conv_hist.append(opt['k_conv'])
+                if k == 0:
```
Comment (jhuggins) [inline viabel/optimization.py:755]: add comment explaining this if/else

File: viabel/optimization.py [hunk 25]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+                                    predicted_iters_hist.append(pred_iters)
+                                relative_iters = pred_iters/(curr_iters + self._iters0)
+                                stopping_crt.append(relative_skl*relative_iters)
+                                if relative_skl*relative_iters > self._inefficiency_threshold:
+                                    stopped = True
+                                    k_stopped_final = k_total
+                                    k_stopped_final_hist.append(k_total)
+                                    break 
+                                
+                    learning_rate_hist.append(sgo._learning_rate)
```
Comment (jhuggins) [inline viabel/optimization.py:855]: why are there two learning rate histories? This seems duplicative and confusing

File: viabel/optimization.py [hunk 26]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+        predicted_iters_hist = []
+        stopping_crt = []
+        k_stopped_final_hist = []
+        LR = [] #learning rate for WLS regression 
+        CI = [] #Converged Iterations for WLS regression
+        try:
+            while not stopped:
+                K_max -= (k_new +1)
+                iterate_average_prev = iterate_average_curr
+                if k==0 and self._init_rmsprop:
```
Comment (jhuggins) [inline viabel/optimization.py:726]: here and elsewhere, please write `k == 0` (with spaces around `==`). Same for, e.g., `k != 0`, etc.

File: viabel/optimization.py [hunk 27]
```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
...
+                    rmsprop = RMSProp(learning_rate=sgo._learning_rate, diagnostics=diagnostics)
+                    faso = FASO(sgo = rmsprop)
+                    opt = faso.optimize(K_max, objective, iterate_average_curr)
+                else:
+                    opt = super().optimize(K_max, objective, iterate_average_curr)
+                if opt['k_stopped'] is not None and k!=0: #removing the number of convergence iterations of initial learning rate
+                    convg_iters = opt['k_stopped']
+                    convg_iters_hist.append(convg_iters)
+                    CI.append(convg_iters)
```
Comment (jhuggins) [inline viabel/optimization.py:735]: same question I had about `LR`. Why the two? Can we avoid the (near?) duplications?

**Review by jhuggins — CHANGES_REQUESTED** [review]

File: viabel/optimization.py [hunk 28]
```diff
@@ -266,40 +290,54 @@ class AveragedRMSProp(StochasticGradientOptimizer):
     descent_dir : `numpy.ndarray`, shape(var_param_dim,)
         Descent direction of the optimization algorithm
     """
-    def __init__(self, learning_rate, *, jitter=1e-8,
-                 diagnostics=False, component_wise=True):
+    def __init__(self, learning_rate, *, beta_1=0.9, beta_2=0.999, jitter=1e-8,
```
Comment (jhuggins) [inline viabel/optimization.py:293]: Please remove the underscores (so just `beta1` and `beta2`) here and below

File: viabel/optimization.py [hunk 29]
```diff
@@ -185,77 +198,88 @@ def descent_direction(self, grad):
...
+    
+    where :math:`\\nu^{(k)}` converges to a constant almost surely under certain 
+    conditions.
     
     Parameters
     -----------
     jitter: `float` optional
         Small value used for numerical stability. The default is 1e-8
+    component_wise: `boolean` optional
+        Indication of  component wise discent direction computation
```
Comment (jhuggins) [inline viabel/optimization.py:222]: Not clear. Do you mean "indicator of whether to do component-wise normalization of descent direction"?

File: viabel/optimization.py [hunk 30]
```diff
@@ -143,12 +147,22 @@ def descent_direction(self, grad):
...
+    and uses :math:`\\nu^{(k)}` to rescale the current stochastic gradient:
+        
+    .. math::
+        \\hat{g}^{(k+1)}/\\sqrt{\\nu^{(k)}}.
     
     Parameters
     -----------
     beta : `float` optional
-        hyper-parameter. The default is 0.9
+        Hyper-parameter. The default is 0.9
```
Comment (jhuggins) [inline viabel/optimization.py:165]: Add something a bit more descriptive here (see comment for Adam)

File: viabel/optimization.py [hunk 31]
```diff
@@ -692,114 +802,56 @@ def optimize(self, K_max, objective, init_param):
...
-                if opt['k_Rhat'] is not None and k_new is not None:
-                    k_Rhat_hist.append(opt['k_Rhat']+k_add)
-                else:
-                    k_Rhat_hist.append(opt['k_Rhat'])
-                    
-                if opt['k_conv'] is not None and k_new is not None:
-                    k_conv_hist.append(opt['k_conv']+k_add)
-                else:
-                    k_conv_hist.append(opt['k_conv'])
+                history['k_Rhat'].append(opt['k_Rhat'] + k_add if opt['k_Rhat'] is not None and k_new is not None else opt['k_Rhat'])
```
Comment (jhuggins) [inline viabel/optimization.py:827]: Why are you checking whether `k_new is not None` here? The logic for this check is not super clear to me so it would be helpful to add a brief comment explaining what's going on

File: viabel/stan_models/weighted_lin_regression.stan [hunk 32]
```diff
@@ -1,4 +1,8 @@
 data {
+    // weighted regression model used to estimate constants that required to
```
Comment (jhuggins) [inline viabel/stan_models/weighted_lin_regression.stan:2]: This comment should go above the data block since it's explaining what the whole model is for

File: viabel/stan_models/weighted_lin_regression_sgd.stan [hunk 33]
```diff
@@ -1,4 +1,8 @@
 data {
+    // weighted regression model used to estimate constants that required to
```
Comment (jhuggins) [inline viabel/stan_models/weighted_lin_regression_sgd.stan:2]: This comment should go above the data block since it's explaining what the whole model is for

**Review by jhuggins — APPROVED** [review]
